lib/src/rules/omit_type_for_local_variables.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.omit_type_for_local_variables;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Omit the types for local variables.';
+
+const _details = r'''
+
+**CONSIDER** omitting the types for local variables.
+
+**BAD:**
+```
+Map<int, List<Person>> groupByZip(Iterable<Person> people) {
+  Map<int, List<Person>> peopleByZip = <int, List<Person>>{};
+  for (Person person in people) {
+    peopleByZip.putIfAbsent(person.zip, () => <Person>[]);
+    peopleByZip[person.zip].add(person);
+  }
+  return peopleByZip;
+}
+```
+
+**GOOD:**
+```
+Map<int, List<Person>> groupByZip(Iterable<Person> people) {
+  var peopleByZip = <int, List<Person>>{};
+  for (var person in people) {
+    peopleByZip.putIfAbsent(person.zip, () => <Person>[]);
+    peopleByZip[person.zip].add(person);
+  }
+  return peopleByZip;
+}
+```
+
+''';
+
+class OmitTypeForLocalVariables extends LintRule {
+  _Visitor _visitor;
+  OmitTypeForLocalVariables()
+      : super(
+            name: 'omit_type_for_local_variables',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitForEachStatement(ForEachStatement node) {
+    final staticType = node.loopVariable.type;
+    if (staticType == null) {
+      return;
+    }
+    if (node.parent is FieldDeclaration) {
+      return;
+    }
+    if (!(node.iterable.bestType is InterfaceType)) {
+      return;
+    }
+    if (DartTypeUtilities.areFromSameClass(
+        (node.iterable.bestType as InterfaceType).typeArguments[0],
+        staticType.type)) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitVariableDeclarationList(VariableDeclarationList node) {
+    final staticType = node.type;
+    if (staticType == null) {
+      return;
+    }
+    if (node.parent is FieldDeclaration) {
+      return;
+    }
+    for (var child in node.variables) {
```

Comment:
I don't know what the coding style rules are exactly for this repo, but in the first line you use `final` but here you use `var`
Doesn't look consistent to me.